dotCMS/src/main/java/com/dotcms/security/secret/ServiceIntegrationAPIImpl.java

```diff
@@ -0,0 +1,451 @@
+                () -> " ymlFiles are set under:  " + ymlFilesPath);
+
+        // Now validate the incoming file.. see if we're rewriting an existing file or attempting to re-use an already in use service-key.
+        final ServiceDescriptor serviceDescriptor = ymlMapper
+                .readValue(inputStream, ServiceDescriptor.class);
+
+        if (validateServiceDescriptor(serviceDescriptor, user)) {
+
+            final String serviceKey = serviceDescriptor.getServiceKey();
+            final File incomingFile = new File(basePath, String.format("%s.yml", serviceKey));
+            if (incomingFile.exists()) {
+                throw new DotDataException(
+                        String.format("Invalid attempt to override an existing file named '%s'.",
+                                incomingFile.getName()));
+            }
+
+            ymlMapper.writeValue(incomingFile, serviceDescriptor);
+
+            invalidateCache();
+        }
+    }
+
+    private void invalidateCache(){
+        CacheLocator
+                .getCacheAdministrator()
+                .remove(DESCRIPTORS_LIST_KEY, DESCRIPTORS_CACHE_GROUP);
+        CacheLocator.getCacheAdministrator()
+                .remove(DESCRIPTORS_MAPPED_BY_SERVICE_KEY, DESCRIPTORS_CACHE_GROUP);
+    }
+
+    private static String getServiceDescriptorDirectory() {
+        final Supplier<String> supplier = () -> APILocator.getFileAssetAPI().getRealAssetsRootPath()
+                + File.separator + "server" + File.separator;
+        final String dirPath = Config
+                .getStringProperty(SERVICE_INTEGRATION_DIR_PATH_KEY, supplier.get());
+        return Paths.get(dirPath).normalize().toString();
+    }
+
+    private Set<String> listAvailableYamlFiles() throws IOException {
+        final String ymlFilesPath = getServiceDescriptorDirectory();
+        final File basePath = new File(ymlFilesPath);
+        if (!basePath.exists()) {
+            basePath.mkdir();
+        }
+        Logger.debug(ServiceIntegrationAPIImpl.class,
+                () -> " ymlFiles are set under:  " + ymlFilesPath);
+        final Set<String> files = listFiles(ymlFilesPath);
+        if (!UtilMethods.isSet(files)) {
+            // No files were found a default empty descriptor will be created.
+            ymlMapper.writeValue(new File(basePath, "sample-service-descriptor.yml"),
+                    emptyDescriptor());
+            return listFiles(ymlFilesPath);
+        } else {
+            return files;
+        }
+    }
+
+    private Set<String> listFiles(final String dir) throws IOException {
+        final Set<String> fileList = new HashSet<>();
+        try (final DirectoryStream<Path> stream = Files.newDirectoryStream(Paths.get(dir))) {
+            for (final Path path : stream) {
+                if (!Files.isDirectory(path)) {
+                    final String fileName = path.getFileName().toString();
+                    if (fileName.equals("yaml") || fileName.endsWith("yml")) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4623512)